bases/renku_data_services/data_api/app.py

```diff
@@ -25,6 +26,9 @@
 
 def register_all_handlers(app: Sanic, config: Config) -> Sanic:
     """Register all handlers on the application."""
+    app.router.register_pattern("ulid", ULID.from_str, r"^[0-9A-HJKMNP-TV-Z]{26}$")
+    app.router.register_pattern("renku_slug", str, r"^[a-zA-Z0-9][a-zA-Z0-9\-_.]*$")
```

Comment:
question: I assume the name (i.e. `ulid`) is the name you define in the route handler right? Like `"/environments/<environment_id:ulid>`? But what is the  difference between `ULID.from_str` and the regex pattern? And I suspect that one of these is run all the time on 100% of the requests that hit the server? I could not understand this from the docs. Do you know?

Also we already have the ULID validation generated from the apispec file. So can't we just leave the regex pattern blank and import the validation from the apispec.py file and use it here? That way we cannot have drift between the apispec and this - because this may cause issues when the apispec and the regex here start to drift.

Lastly a valid ulid does not contain the letters `I,L,O,U`? I did not know this.

Reply:
yes `ulid` is the name, the second parameter is the function sanic uses to convert to the target type and the regex is what is uses to match routes (so not validation really but just how sanic knows this is the route to use). So the regex is required. Note that this is sort of a hack as the regex is for route matching, not validation. The "proper" way to do it would be to add a line of code to validate these things in every handler and raise a validation exception, but this seemed cleaner/easier to me and I at least could make an argument that this is correct behavior :slightly_smiling_face: . But it does raise a 404 route not found instead of a validation error if something doesn't match. That's enough to make schemathesis happy, and I don't really want to add something like
```
try:
    ULID.from_str(id)
except ValueError:
   raise ValidationError("Invalid ULID")
```
in every handler that takes an ULID.

The APIspec only validates post bodies, we (or rather, sanic) don't use the apispec for path parameters, hence this change. Schemathesis otherwise fails because `DELETE environments/a` should be an error but passes (due to "not found" not being an error in the DB layer). We could reference one of the APIspec directly, though we already have 7 ULID definitions in different apispecs, so what's an 8th one? :shrug: ULID probably won't change anytime soon, slug is probably the bigger issue. That one we have in 2 places so far. Would be nice if we could have a single place that we include here AND in the apispec.

ULID uses https://en.wikipedia.org/wiki/Base32#Crockford's_Base32 , which doesn't include I, L, O to avoid confusion with numbers, and U to not have profanity in the Id :rofl: (Can't have f*ck in an ULID!)

Reply:
Pedantic note: removing `U` also avoids confusion with `V` when using hand writing (same as `I`, `L` and `O`).

Reply:
pedantic reply: True, but profanity is the reason that wikipedia mentions for dropping the U :man_shrugging: